File: go/modcdp/transport/NatsUpstreamTransport.go
package transport

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"net"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

const DefaultUpstreamNATSURL = "ws://127.0.0.1:4223"
const DefaultUpstreamNATSSubjectPrefix = "modcdp.default"
const DefaultUpstreamNATSWaitTimeoutMS = 10_000

type NatsUpstreamTransport struct {
	UpstreamTransport
	URL                       string
	UpstreamNATSSubjectPrefix string
	UpstreamNATSRole          string
	WaitTimeoutMS             int
	Conn                      net.Conn
	IsWebSocket               bool
	buffer                    string
	connected                 bool
	closed                    bool
	writeMu                   sync.Mutex
	peerCh                    chan struct{}
	peerOnce                  sync.Once
	closeCh                   chan struct{}
	stateMu                   sync.Mutex
}

type NatsUpstreamTransportOptions struct {
	UpstreamNATSURL           string `json:"upstream_nats_url,omitempty"`
	UpstreamNATSSubjectPrefix string `json:"upstream_nats_subject_prefix,omitempty"`
	UpstreamNATSRole          string `json:"upstream_nats_role,omitempty"`
	UpstreamNATSWaitTimeoutMS int    `json:"upstream_nats_wait_timeout_ms,omitempty"`
}

func NewNatsUpstreamTransport(options NatsUpstreamTransportOptions) *NatsUpstreamTransport {
	normalizedURL, subjectPrefix := normalizeUpstreamNATSURL(firstNonEmptyString(options.UpstreamNATSURL, DefaultUpstreamNATSURL), options.UpstreamNATSSubjectPrefix)
	role := firstNonEmptyString(options.UpstreamNATSRole, "client")
	waitTimeoutMS := options.UpstreamNATSWaitTimeoutMS
	if waitTimeoutMS == 0 {
		waitTimeoutMS = DefaultUpstreamNATSWaitTimeoutMS
	}
	return &NatsUpstreamTransport{
		URL:                       normalizedURL,
		UpstreamNATSSubjectPrefix: subjectPrefix,
		UpstreamNATSRole:          role,
		WaitTimeoutMS:             waitTimeoutMS,
		peerCh:                    make(chan struct{}),
		closeCh:                   make(chan struct{}),
	}
}

func (t *NatsUpstreamTransport) Update(config map[string]any) {
	if config == nil {
		return
	}
	natsURL, _ := config["upstream_nats_url"].(string)
	subjectPrefix, _ := config["upstream_nats_subject_prefix"].(string)
	if natsURL != "" || subjectPrefix != "" {
		t.URL, t.UpstreamNATSSubjectPrefix = normalizeUpstreamNATSURL(firstNonEmptyString(natsURL, t.URL), firstNonEmptyString(subjectPrefix, t.UpstreamNATSSubjectPrefix))
	}
	if role, _ := config["upstream_nats_role"].(string); role == "client" || role == "browser" {
		t.UpstreamNATSRole = role
	}
	if waitTimeoutMS, ok := intFromConfig(config["upstream_nats_wait_timeout_ms"]); ok {
		t.WaitTimeoutMS = waitTimeoutMS
	}
}

func (t *NatsUpstreamTransport) GetInjectorConfig() ExtensionInjectorConfig {
	return ExtensionInjectorConfig{UpstreamNATSURL: t.URL, UpstreamNATSSubjectPrefix: t.UpstreamNATSSubjectPrefix}
}

func (t *NatsUpstreamTransport) Connect() error {
	if t.connected {
		return nil
	}
	t.stateMu.Lock()
	t.closed = false
	t.closeCh = make(chan struct{})
	closeCh := t.closeCh
	t.stateMu.Unlock()
	if !validUpstreamNATSSubjectPrefix(t.UpstreamNATSSubjectPrefix) {
		return fmt.Errorf("invalid NATS subject prefix %q", t.UpstreamNATSSubjectPrefix)
	}
	parsed, err := url.Parse(t.URL)
	if err != nil {
		return err
	}
	switch parsed.Scheme {
	case "ws", "wss":
		conn, _, _, err := ws.Dial(context.Background(), t.URL)
		if err != nil {
			return err
		}
		t.Conn = conn
		t.IsWebSocket = true
		if err := t.writeProtocol("CONNECT " + mustJSON(connectNATSOptions()) + "\r\nPING\r\n"); err != nil {
			_ = conn.Close()
			return err
		}
		go t.readWebSocketLoop(conn, closeCh)
	case "nats", "tls":
		host := parsed.Hostname()
		if host == "" {
			host = "127.0.0.1"
		}
		port := parsed.Port()
		if port == "" {
			port = "4222"
		}
		var conn net.Conn
		if parsed.Scheme == "tls" {
			conn, err = tls.Dial("tcp", net.JoinHostPort(host, port), &tls.Config{ServerName: host})
		} else {
			conn, err = net.Dial("tcp", net.JoinHostPort(host, port))
		}
		if err != nil {
			return err
		}
		t.Conn = conn
		if err := t.writeProtocol("CONNECT " + mustJSON(connectNATSOptions()) + "\r\nPING\r\n"); err != nil {
			_ = conn.Close()
			return err
		}
		go t.readTCPLoop(conn, closeCh)
	default:
		return fmt.Errorf("upstream.mode=nats requires ws://, wss://, nats://, or tls:// URL, got %s", t.URL)
	}
	t.connected = true
	if err := t.subscribe(); err != nil {
		return err
	}
	return t.publish(t.outgoingSubject(), map[string]any{"type": "modcdp.nats.hello", "role": t.UpstreamNATSRole, "version": 1})
}

func (t *NatsUpstreamTransport) Send(message map[string]any) error {
	if !t.connected || t.Conn == nil {
		return fmt.Errorf("NATS transport is not connected")
	}
	return t.publish(t.outgoingSubject(), map[string]any{"type": "modcdp.nats.message", "message": message})
}

func (t *NatsUpstreamTransport) WaitForPeer() error {
	t.stateMu.Lock()
	closeCh := t.closeCh
	t.stateMu.Unlock()
	select {
	case <-t.peerCh:
		return nil
	case <-closeCh:
		return fmt.Errorf("NATS transport for %s closed before a peer connected", t.UpstreamNATSSubjectPrefix)
	case <-time.After(time.Duration(t.WaitTimeoutMS) * time.Millisecond):
		return fmt.Errorf("timed out waiting %dms for NATS ModCDP peer", t.WaitTimeoutMS)
	}
}

func (t *NatsUpstreamTransport) Close() error {
	t.stateMu.Lock()
	t.closed = true
	closeCh := t.closeCh
	t.closeCh = make(chan struct{})
	t.stateMu.Unlock()
	close(closeCh)
	t.writeMu.Lock()
	t.connected = false
	if t.Conn != nil {
		_ = t.Conn.Close()
		t.Conn = nil
	}
	t.writeMu.Unlock()
	t.buffer = ""
	t.peerCh = make(chan struct{})
	t.peerOnce = sync.Once{}
	return nil
}

func (t *NatsUpstreamTransport) Connected() bool {
	return t.connected
}

func (t *NatsUpstreamTransport) Closed() bool {
	return t.closed
}

func natsClosed(closeCh chan struct{}) bool {
	select {
	case <-closeCh:
		return true
	default:
		return false
	}
}

func (t *NatsUpstreamTransport) subscribe() error {
	return t.writeProtocol("SUB " + t.incomingSubject() + " 1\r\n")
}

func (t *NatsUpstreamTransport) publish(subject string, message map[string]any) error {
	body, err := json.Marshal(message)
	if err != nil {
		return err
	}
	return t.writeProtocol(fmt.Sprintf("PUB %s %d\r\n%s\r\n", subject, len(body), string(body)))
}

func (t *NatsUpstreamTransport) writeProtocol(data string) error {
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	conn := t.Conn
	if conn == nil {
		return fmt.Errorf("NATS transport is not connected")
	}
	if t.IsWebSocket {
		return wsutil.WriteClientText(conn, []byte(data))
	}
	_, err := conn.Write([]byte(data))
	return err
}

func (t *NatsUpstreamTransport) incomingSubject() string {
	if t.UpstreamNATSRole == "client" {
		return t.UpstreamNATSSubjectPrefix + ".browser_to_client"
	}
	return t.UpstreamNATSSubjectPrefix + ".client_to_browser"
}

func (t *NatsUpstreamTransport) outgoingSubject() string {
	if t.UpstreamNATSRole == "client" {
		return t.UpstreamNATSSubjectPrefix + ".client_to_browser"
	}
	return t.UpstreamNATSSubjectPrefix + ".browser_to_client"
}

func (t *NatsUpstreamTransport) readWebSocketLoop(conn net.Conn, closeCh chan struct{}) {
	for !natsClosed(closeCh) {
		data, _, err := wsutil.ReadServerData(conn)
		if err != nil {
			if !natsClosed(closeCh) {
				t.emitClose(err)
			}
			return
		}
		t.buffer = t.consumeProtocol(t.buffer + string(data))
	}
}

func (t *NatsUpstreamTransport) readTCPLoop(conn net.Conn, closeCh chan struct{}) {
	chunk := make([]byte, 65536)
	for !natsClosed(closeCh) {
		n, err := conn.Read(chunk)
		if err != nil {
			if !natsClosed(closeCh) {
				t.emitClose(err)
			}
			return
		}
		t.buffer = t.consumeProtocol(t.buffer + string(chunk[:n]))
	}
}

func (t *NatsUpstreamTransport) consumeProtocol(buffer string) string {
	for {
		lineEnd := strings.Index(buffer, "\r\n")
		if lineEnd < 0 {
			return buffer
		}
		line := buffer[:lineEnd]
		upper := strings.ToUpper(line)
		if strings.HasPrefix(upper, "MSG ") {
			parts := strings.Fields(line)
			size, err := strconv.Atoi(parts[len(parts)-1])
			payloadStart := lineEnd + 2
			payloadEnd := payloadStart + size
			if err != nil || len(buffer) < payloadEnd+2 {
				return buffer
			}
			payload := buffer[payloadStart:payloadEnd]
			buffer = buffer[payloadEnd+2:]
			t.handlePayload(payload)
			continue
		}
		buffer = buffer[lineEnd+2:]
		if upper == "PING" {
			_ = t.writeProtocol("PONG\r\n")
		} else if strings.HasPrefix(upper, "-ERR") {
			t.emitClose(fmt.Errorf("NATS error: %s", line))
		}
	}
}

func (t *NatsUpstreamTransport) handlePayload(payload string) {
	var parsed any
	if err := json.Unmarshal([]byte(payload), &parsed); err != nil {
		return
	}
	record, _ := parsed.(map[string]any)
	if record["type"] == "modcdp.nats.hello" {
		t.peerOnce.Do(func() { close(t.peerCh) })
		return
	}
	message := parsed
	if record["type"] == "modcdp.nats.message" {
		message = record["message"]
	}
	if cdpMessage, ok := message.(map[string]any); ok {
		t.emitRecv(cdpMessage)
	}
}

func (t *NatsUpstreamTransport) HandlePayload(payload string) {
	t.handlePayload(payload)
}

func connectNATSOptions() map[string]any {
	return map[string]any{"verbose": false, "pedantic": false, "lang": "modcdp", "version": "1", "protocol": 1}
}

func normalizeUpstreamNATSURL(rawURL string, subjectPrefix string) (string, string) {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return rawURL, sanitizeUpstreamNATSSubjectPrefix(firstNonEmptyString(subjectPrefix, DefaultUpstreamNATSSubjectPrefix))
	}
	query := parsed.Query()
	subject := firstNonEmptyString(subjectPrefix, query.Get("upstream_nats_subject_prefix"), DefaultUpstreamNATSSubjectPrefix)
	query.Del("upstream_nats_subject_prefix")
	parsed.RawQuery = query.Encode()
	if parsed.Path == "" && (parsed.Scheme == "ws" || parsed.Scheme == "wss") {
		parsed.Path = "/"
	}
	return parsed.String(), sanitizeUpstreamNATSSubjectPrefix(subject)
}

func sanitizeUpstreamNATSSubjectPrefix(value string) string {
	return strings.TrimSpace(value)
}

func validUpstreamNATSSubjectPrefix(value string) bool {
	subject := strings.TrimSpace(value)
	return subject != "" && !strings.ContainsAny(subject, " \t\r\n*>")
}

func mustJSON(value any) string {
	body, _ := json.Marshal(value)
	return string(body)
}

File: python/modcdp/transport/NatsUpstreamTransport.py
from __future__ import annotations

import json
import socket
import ssl
import threading
import time
from collections.abc import Mapping
from typing import Any, cast
from urllib.parse import parse_qs, urlencode, urlparse, urlunparse

from websocket import WebSocket

from ..transport.UpstreamTransport import UpstreamTransport


DEFAULT_UPSTREAM_NATS_URL = "ws://127.0.0.1:4223"
DEFAULT_UPSTREAM_NATS_SUBJECT_PREFIX = "modcdp.default"
DEFAULT_UPSTREAM_NATS_WAIT_TIMEOUT_MS = 10_000


class NatsUpstreamTransport(UpstreamTransport):
    mode = "nats"
    endpoint_kind = "modcdp_server"

    def __init__(
        self,
        options: Mapping[str, Any] | None = None,
    ) -> None:
        super().__init__()
        normalized_options = dict(options or {})
        normalized_url, normalized_nats_subject_prefix = _normalize_nats_url(
            cast(str | None, normalized_options.get("upstream_nats_url")) or DEFAULT_UPSTREAM_NATS_URL,
            cast(str | None, normalized_options.get("upstream_nats_subject_prefix")),
        )
        self.url = normalized_url
        self.upstream_nats_subject_prefix = normalized_nats_subject_prefix
        self.upstream_nats_role = str(normalized_options.get("upstream_nats_role") or "client")
        self.wait_timeout_ms = int(normalized_options.get("upstream_nats_wait_timeout_ms") or DEFAULT_UPSTREAM_NATS_WAIT_TIMEOUT_MS)
        self.socket: WebSocket | socket.socket | None = None
        self.connected = False
        self.closed = False
        self.peer_seen = threading.Event()
        self._peer_condition = threading.Condition()
        self._close_generation = 0
        self.write_lock = threading.Lock()
        self.buffer = ""

    def update(self, config: dict[str, Any] | None = None) -> "NatsUpstreamTransport":
        config = config or {}
        upstream_nats_url = config.get("upstream_nats_url")
        upstream_nats_subject_prefix = config.get("upstream_nats_subject_prefix")
        if isinstance(upstream_nats_url, str) and upstream_nats_url or isinstance(upstream_nats_subject_prefix, str) and upstream_nats_subject_prefix:
            current_url = self.url or DEFAULT_UPSTREAM_NATS_URL
            self.url, self.upstream_nats_subject_prefix = _normalize_nats_url(
                upstream_nats_url if isinstance(upstream_nats_url, str) and upstream_nats_url else current_url,
                upstream_nats_subject_prefix if isinstance(upstream_nats_subject_prefix, str) and upstream_nats_subject_prefix else self.upstream_nats_subject_prefix,
            )
        upstream_nats_role = config.get("upstream_nats_role")
        if upstream_nats_role in ("client", "browser"):
            self.upstream_nats_role = str(upstream_nats_role)
        wait_timeout_ms = config.get("upstream_nats_wait_timeout_ms")
        if isinstance(wait_timeout_ms, int | float):
            self.wait_timeout_ms = int(wait_timeout_ms)
        return self

    def getInjectorConfig(self) -> dict[str, Any]:
        return {"upstream_nats_url": self.url, "upstream_nats_subject_prefix": self.upstream_nats_subject_prefix}

    def connect(self) -> None:
        if self.connected:
            return
        self.closed = False
        with self._peer_condition:
            self._close_generation += 1
            close_generation = self._close_generation
            self.peer_seen.clear()
        parsed = urlparse(self.url)
        if parsed.scheme in ("ws", "wss"):
            ws = WebSocket()
            ws.connect(self.url)
            self.socket = ws
            self._write_protocol(f"CONNECT {json.dumps(_connect_options())}\r\nPING\r\n")
            threading.Thread(target=self._read_websocket_loop, args=(ws, close_generation), daemon=True).start()
        elif parsed.scheme in ("nats", "tls"):
            port = parsed.port or 4222
            host = cast(str, parsed.hostname or "127.0.0.1")
            raw_socket = socket.create_connection((host, port))
            tcp_socket = ssl.create_default_context().wrap_socket(raw_socket, server_hostname=host) if parsed.scheme == "tls" else raw_socket
            self.socket = tcp_socket
            self._write_protocol(f"CONNECT {json.dumps(_connect_options())}\r\nPING\r\n")
            threading.Thread(target=self._read_tcp_loop, args=(tcp_socket, close_generation), daemon=True).start()
        else:
            raise RuntimeError(f"upstream.mode=nats requires ws://, wss://, nats://, or tls:// URL, got {self.url}.")
        self.connected = True
        self._subscribe()
        self._publish(self._outgoing_subject(), {"type": "modcdp.nats.hello", "role": self.upstream_nats_role, "version": 1})

    def send(self, message: dict[str, Any]) -> None:
        if not self.connected or self.socket is None:
            raise RuntimeError("NATS transport is not connected.")
        self._publish(self._outgoing_subject(), {"type": "modcdp.nats.message", "message": message})

    def waitForPeer(self) -> None:
        deadline = time.monotonic() + self.wait_timeout_ms / 1000
        with self._peer_condition:
            close_generation = self._close_generation
            while not self.peer_seen.is_set():
                if close_generation != self._close_generation:
                    raise RuntimeError(f"NATS transport for {self.upstream_nats_subject_prefix} closed before a peer connected.")
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise RuntimeError(f"Timed out waiting {self.wait_timeout_ms}ms for NATS ModCDP peer.")
                self._peer_condition.wait(remaining)

    def close(self) -> None:
        self.closed = True
        try:
            if isinstance(self.socket, WebSocket):
                self.socket.close()
            elif self.socket is not None:
                self.socket.close()
        except Exception:
            pass
        self.socket = None
        self.connected = False
        self.peer_seen.clear()
        with self._peer_condition:
            self._close_generation += 1
            self._peer_condition.notify_all()

    def _subscribe(self) -> None:
        self._write_protocol(f"SUB {self._incoming_subject()} 1\r\n")

    def _publish(self, subject: str, message: dict[str, Any]) -> None:
        body = json.dumps(message, separators=(",", ":"))
        self._write_protocol(f"PUB {subject} {len(body.encode())}\r\n{body}\r\n")

    def _write_protocol(self, data: str) -> None:
        if self.socket is None:
            raise RuntimeError("NATS transport is not connected.")
        with self.write_lock:
            if isinstance(self.socket, WebSocket):
                self.socket.send(data)
            else:
                self.socket.sendall(data.encode())

    def _incoming_subject(self) -> str:
        return f"{self.upstream_nats_subject_prefix}.{'browser_to_client' if self.upstream_nats_role == 'client' else 'client_to_browser'}"

    def _outgoing_subject(self) -> str:
        return f"{self.upstream_nats_subject_prefix}.{'client_to_browser' if self.upstream_nats_role == 'client' else 'browser_to_client'}"

    def _generation_closed(self, close_generation: int) -> bool:
        with self._peer_condition:
            return self.closed or close_generation != self._close_generation

    def _read_websocket_loop(self, ws: WebSocket, close_generation: int) -> None:
        try:
            while not self._generation_closed(close_generation):
                data = ws.recv()
                if isinstance(data, bytes):
                    self.buffer += data.decode()
                else:
                    self.buffer += str(data)
                self.buffer = self._consume_protocol(self.buffer)
        except Exception as error:
            if not self._generation_closed(close_generation):
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))

    def _read_tcp_loop(self, tcp_socket: socket.socket, close_generation: int) -> None:
        try:
            while not self._generation_closed(close_generation):
                chunk = tcp_socket.recv(65536)
                if not chunk:
                    break
                self.buffer += chunk.decode()
                self.buffer = self._consume_protocol(self.buffer)
        except Exception as error:
            if not self._generation_closed(close_generation):
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))

    def _consume_protocol(self, buffer: str) -> str:
        while True:
            line_end = buffer.find("\r\n")
            if line_end < 0:
                return buffer
            line = buffer[:line_end]
            upper = line.upper()
            if upper.startswith("MSG "):
                parts = line.split()
                size = int(parts[-1]) if parts and parts[-1].isdigit() else -1
                payload_start = line_end + 2
                payload_end = payload_start + size
                if size < 0 or len(buffer) < payload_end + 2:
                    return buffer
                payload = buffer[payload_start:payload_end]
                buffer = buffer[payload_end + 2 :]
                self._handle_payload(payload)
                continue
            buffer = buffer[line_end + 2 :]
            if upper == "PING":
                self._write_protocol("PONG\r\n")
            elif upper.startswith("-ERR"):
                self._emit_close(RuntimeError(f"NATS error: {line}"))

    def _handle_payload(self, payload: str) -> None:
        try:
            parsed = json.loads(payload)
        except Exception:
            return
        if isinstance(parsed, dict) and parsed.get("type") == "modcdp.nats.hello":
            with self._peer_condition:
                self.peer_seen.set()
                self._peer_condition.notify_all()
            return
        message = parsed.get("message") if isinstance(parsed, dict) and parsed.get("type") == "modcdp.nats.message" else parsed
        if isinstance(message, dict):
            self._emit_recv(message)


def _connect_options() -> dict[str, Any]:
    return {"verbose": False, "pedantic": False, "lang": "modcdp", "version": "1", "protocol": 1}


def _normalize_nats_url(url: str, upstream_nats_subject_prefix: str | None = None) -> tuple[str, str]:
    parsed = urlparse(url)
    query = parse_qs(parsed.query)
    subject = upstream_nats_subject_prefix or (query.get("upstream_nats_subject_prefix") or [None])[0]
    query.pop("upstream_nats_subject_prefix", None)
    normalized_query = urlencode(query, doseq=True)
    normalized_path = parsed.path or ("/" if parsed.scheme in ("ws", "wss") else "")
    normalized_url = urlunparse((parsed.scheme, parsed.netloc, normalized_path, parsed.params, normalized_query, parsed.fragment))
    return normalized_url, _sanitize_nats_subject_prefix(subject or DEFAULT_UPSTREAM_NATS_SUBJECT_PREFIX)


def _sanitize_nats_subject_prefix(value: str) -> str:
    subject = value.strip()
    if not subject or any(char.isspace() or char in "*>" for char in subject):
        raise ValueError(f"Invalid NATS subject prefix {value}")
    return subject
